### PR TITLE
[dev] Update pango, harfbuzz, glib, python3, llvm and add fribidi

### PR DESCRIPTION
**Pull request by henryli001**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?

Certain package has dependency on a higher version of pango than the one currently supported in Mariner. The newer version of pango requires build requirement on fribidi and also a higher version of glib and harfbuzz than the ones currently in Mariner. Therefore, this PR will add fribidi to the repo and upgrade the version of glib, pango and harfbuzz.

Some packages have build requirement on python without mentioning the version (i.e. python2 vs python3). Thus, add provide for python from python3 in order to satisfy that requirement.

Specific package that depends on llvm requires the existence of rtti instances in order to compile. Thus, adding enable rtti to the build option to satisfy the compilation requirement.

###### Change Log  

1. glib:
- Upgrade to version 2.60.1
- Switch to meson build and install 
- Fix file section for glib-devel

2. harfbuzz:
- Upgrade to version 2.6.4

3.pango:
- Upgrade to version 1.44.7
- Switch to meson build and install
- Add meson and pkgconfig(fribidi) as build requirement
- Fix file section for pango

4. python3:
- Provides python from python3

5. llvm:
-Add -DLLVM_ENABLE_RTTI=ON to cmake build option

###### Does this affect the toolchain?  

YES

###### Associated issues  

- #xxxx

###### Links to CVEs  
- https://nvd.nist.gov/vuln/detail/CVE-YYYY-XXXX

###### Test Methodology

- Local Build

### CODE REVIEW

File: SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md [hunk 1]
```diff
@@ -2,7 +2,7 @@ The CBL-Mariner SPEC files originated from a variety of sources with varying lic
...
...br> Cython <br> ant-contrib <br> atf <br> attr <br> autoconf213 <br> babeltrace2 <br> brotli <br> byacc <br> calamares <br> cairo <br> catch <br> cpprest <br> cryptsetup <br> dnf <br> dpdk <br> docbook5-schemas <br> dos2unix <br> dwz <br> ebtables <br> extra-cmake-modules <br> fribidi <br> gnu-efi <br> hardening-check <br> ipmitool <br> javapackages-tools <br> kde-settings <br> kexec-tools <br> kf5 <br> kf5-kconfig <br> kf5-kcoreaddons <br> kf5-ki18n <br> kf5-kwidgetsaddons <br> kpmcore-3.3.0 <br> kyua <br> leveldb <br> libatasmart <br> libburn <br> libcgroup <br> libcomps <br> libdnf <br> libisoburn <br> libisofs <br> libpwquality <br> libstoragemgmt <br> lua <br> lutok <br> mailcap <br> mokutil <br> mozjs60 <br> oath-toolkit <br> open-vm-tools <br> openssl <br> p11-kit <br> perl <br> perl-Archive-Zip <br> perl-Compress-Bzip2 <br> pkgconf <br> pycairo <br> pygobject3 <br> python-google-auth <br> python3 <br> python-jwt <br> python-kubernetes <br> python-pywbem <br> qt5-qtbase <br> qt5-qtdeclarative <br> qt5-qtsvg <br> qt5-qttools <br> qt5-rpm-macros <br> rdma-core <br> sgml-common <br> squashfs-tools <br> websocketpp <br> words <br> xmlstarlet <br> yaml-cpp <br> yasm <br> zipper |
```
Comment (PawelWMS) [inline SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md:5]: Could you also mention in the PR's summary why is we're adding/upgrading these specs. I'm not saying we shouldn't, it is just shrouded in mystery at the moment.;)

File: SPECS/fribidi/fribidi.spec [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+%files devel
+%{_includedir}/fribidi
+%{_libdir}/libfribidi.so
+%{_libdir}/pkgconfig/*.pc
+#%%{_mandir}/man3/*.gz
+
+%changelog
+* Thu Apr 15 2021 Henry Li <lihl@microsoft.com> - 1.0.9-2
+- Initial import from Fedora 32 (license: MIT)
```
Comment (PawelWMS) [inline SPECS/fribidi/fribidi.spec:57]: Please follow the format from other specs, which is "Initial CBL-Mariner import from Fedora 32 (license: MIT).".

File: SPECS/fribidi/fribidi.spec [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+- Remove distro check that does not apply to CBL-Mariner
```
Comment (PawelWMS) [inline SPECS/fribidi/fribidi.spec:58]: Please check the license, update the `License` tag and the `%license` macro (if needed), and add the phrase `License verified` to the changelog once that's done.

File: SPECS/fribidi/fribidi.spec [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+%doc README AUTHORS ChangeLog THANKS NEWS TODO
+%license COPYING
+%{_bindir}/fribidi
+%{_libdir}/libfribidi.so.0*
+
+%files devel
+%{_includedir}/fribidi
+%{_libdir}/libfribidi.so
+%{_libdir}/pkgconfig/*.pc
+#%%{_mandir}/man3/*.gz
```
Comment (PawelWMS) [inline SPECS/fribidi/fribidi.spec:53]: Please remove commented code.

File: SPECS/glib/glib.spec [hunk 5]
```diff
@@ -69,18 +71,19 @@ The glib2-doc package includes documentation for the GLib library.
...
-%patch0 -p1
-%patch1 -p1
 
 %build
-./autogen.sh
-%configure --with-pcre=system \
-            --enable-static \
-            --enable-gtk-doc
-make %{?_smp_mflags}
+%meson \
```
Comment (PawelWMS) [inline SPECS/glib/glib.spec:76]: Double-checking: are we still keeping the equivalent of the `--with-pcre` switch from the earlier version?

File: SPECS/glib/glib.spec [hunk 6]
```diff
@@ -99,7 +102,6 @@ make DESTDIR=%{buildroot} install
 %{_bindir}/*
 %{_libdir}/*.a
 %{_libdir}/*.so
-%{_libdir}/*.la
```
Comment (PawelWMS) [inline SPECS/glib/glib.spec:102]: Why is this file gone now? I don't think anyone needs it, just curious.

File: SPECS/glib/glib.spec [hunk 7]
```diff
@@ -69,18 +71,19 @@ The glib2-doc package includes documentation for the GLib library.
...
+    --default-library=both
+
+%meson_build
 
 %install
-make DESTDIR=%{buildroot} install
+%meson_install
+
+mkdir -p %{buildroot}%{_libdir}/gio/modules
+touch %{buildroot}%{_libdir}/gio/modules/giomodule.cache
```
Comment (PawelWMS) [inline SPECS/glib/glib.spec:86]: Why do we create this manually now? What was it used for in the previous versions and is it OK for this file to be empty?

File: SPECS/llvm/llvm.spec [hunk 8]
```diff
@@ -37,8 +37,9 @@ export CXXFLAGS="`echo " %{build_cxxflags} " | sed 's/ -g//'`"
 mkdir -p build
 cd build
 cmake -G Ninja                              \
-      -DCMAKE_INSTALL_PREFIX=/usr           \
+      -DCMAKE_INSTALL_PREFIX=%{_prefix}           \
```
Comment (PawelWMS) [inline SPECS/llvm/llvm.spec:40]: Nit: could you also re-align the `\` signs, since they were so neatly done before? ;)

File: SPECS/llvm/llvm.spec [hunk 9]
```diff
@@ -37,8 +37,9 @@ export CXXFLAGS="`echo " %{build_cxxflags} " | sed 's/ -g//'`"
...
       -DLLVM_ENABLE_FFI=ON                  \
+      -DLLVM_ENABLE_RTTI=ON                 \
```
Comment (PawelWMS) [inline SPECS/llvm/llvm.spec:42]: Please mention why we're making this change in the PR's summary.

File: SPECS/pango/pango.spec [hunk 10]
```diff
@@ -1,58 +1,68 @@
...
-URL:		http://pango.org
-Group:		System Environment/Libraries
+Summary:        library for laying out and rendering of text.
+Name:           pango
+Version:        1.44.7
+Release:        1%{?dist}
+License:        LGPLv2 OR MPLv1.1
 Vendor:         Microsoft Corporation
 Distribution:   Mariner
-Source0:	    https://download.gnome.org/sources/pango/1.40/%{name}-%{version}.tar.xz
```
Comment (PawelWMS) [inline SPECS/pango/pango.spec:10]: Thanks for running the linter in the specs!

File: SPECS/pango/pango.spec [hunk 11]
```diff
@@ -1,58 +1,68 @@
...
+BuildRequires:  fontconfig-devel
+BuildRequires:  freetype
+BuildRequires:  glib-devel
+BuildRequires:  gobject-introspection-devel
+BuildRequires:  harfbuzz
+BuildRequires:  harfbuzz-devel
+BuildRequires:  libpng-devel
+BuildRequires:  meson
+BuildRequires:  pkg-config
+BuildRequires:  pkgconfig(fribidi)
```
Comment (PawelWMS) [inline SPECS/pango/pango.spec:23]: Why is this required to build now? I'm curious if it's related to the new `*.typelib` files being produced and maybe we can disable their generation along with a dependency on `pkgconfig(fribidi)` in one go.

File: SPECS/glib/glib.spec (hunk 5 above)
Comment (henryli001) [reply SPECS/glib/glib.spec:76]: Yes, the new version of glib is using the system version of pcre by default

File: SPECS/glib/glib.spec (hunk 6 above)
Comment (henryli001) [reply SPECS/glib/glib.spec:102]: I believe this is due to the change of the implementation of different versions of glib. In the previous version, I can see the libtool libraries being defined in the Makefile, whereas in the newer version, they no longer exist and thus no *.la files will be created

File: SPECS/pango/pango.spec (hunk 11 above)
Comment (henryli001) [reply SPECS/pango/pango.spec:23]: This is also a change due to the upgrade of the pango version. In the meson.build file of the new pango source code, it lists fribidi as a dependency for pango and the version of fribidi that pango is looking for.

File: SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md (hunk 1 above)
Comment (henryli001) [reply SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md:5]: Have mentioned the reason

File: SPECS/fribidi/fribidi.spec (hunk 3 above)
Comment (henryli001) [reply SPECS/fribidi/fribidi.spec]: Fixed

File: SPECS/fribidi/fribidi.spec (hunk 4 above)
Comment (henryli001) [reply SPECS/fribidi/fribidi.spec]: Fixed

File: SPECS/fribidi/fribidi.spec (hunk 2 above)
Comment (henryli001) [reply SPECS/fribidi/fribidi.spec]: Fixed

File: SPECS/glib/glib.spec (hunk 7 above)
Comment (henryli001) [reply SPECS/glib/glib.spec:89]: In the configuration file of the previous version, there's a section of code that will determine the default path of gio modules directory and that piece of code is gone in the newer version. In the glib spec file of other distributions, it's also manually creating this directory.

In the previous version, we don't even have this cache file. The gio/modules directory is empty.

**Review by jslobodzian — APPROVED** [review]

File: SPECS/glib/glib.spec [hunk 12]
```diff
@@ -1,23 +1,25 @@
 Summary:        Low-level libraries useful for providing data structure handling for C.
 Name:           glib
-Version:        2.58.0
-Release:        12%{?dist}
+Version:        2.60.1
+Release:        1%{?dist}
 License:        LGPLv2+
 Vendor:         Microsoft Corporation
 Distribution:   Mariner
 Group:          Applications/System
```
Comment (MateuszMalisz) [inline SPECS/glib/glib.spec:8]: Nit: since groups are deprecated and do not bring any value, I suggest removing this entry.

File: SPECS/harfbuzz/harfbuzz.spec [hunk 13]
```diff
@@ -1,46 +1,45 @@
...
+Release:        1%{?dist}
+License:        MIT
 Vendor:         Microsoft Corporation
 Distribution:   Mariner
-Source0:	    https://www.freedesktop.org/software/harfbuzz/release/%{name}-%{version}.tar.bz2
-BuildRequires:	glib-devel
-BuildRequires:	freetype
-BuildRequires:	freetype-devel
-Requires:	    glib
+Group:          System Environment/Libraries
```
Comment (MateuszMalisz) [inline SPECS/harfbuzz/harfbuzz.spec:8]: Same comment as the above regarding the group

File: SPECS/llvm/llvm.spec [hunk 14]
```diff
@@ -1,18 +1,18 @@
...
 Version:        8.0.1
-Release:        3%{?dist}
+Release:        4%{?dist}
 License:        NCSA
-URL:            https://llvm.org/
-Source0:        https://github.com/llvm/llvm-project/releases/download/llvmorg-%{version}/%{name}-%{version}.src.tar.xz
-Group:          Development/Tools
 Vendor:         Microsoft Corporation
 Distribution:   Mariner
+Group:          Development/Tools
```
Comment (MateuszMalisz) [inline SPECS/llvm/llvm.spec:8]: Same comment as above regarding the group

File: SPECS/pango/pango.spec [hunk 15]
```diff
@@ -1,58 +1,68 @@
...
-BuildRequires:	fontconfig
-BuildRequires:	fontconfig-devel
-BuildRequires:	harfbuzz
-BuildRequires:	harfbuzz-devel
-BuildRequires:	freetype
-Requires:	    harfbuzz-devel
+Group:          System Environment/Libraries
+URL:            https://pango.org
+Source0:        https://download.gnome.org/sources/pango/1.44/%{name}-%{version}.tar.xz
+BuildRequires:  cairo
```
Comment (MateuszMalisz) [inline SPECS/pango/pango.spec:11]: Can we drop the base packages? It seems that `-devel` packages depend on the base ones already.

File: SPECS/harfbuzz/harfbuzz.spec [hunk 16]
```diff
@@ -1,46 +1,45 @@
...
+URL:            http://harfbuzz.org
+Source0:        https://www.freedesktop.org/software/harfbuzz/release/%{name}-%{version}.tar.xz
+BuildRequires:  freetype
```
Comment (MateuszMalisz) [inline SPECS/harfbuzz/harfbuzz.spec:11]: Can we drop the base packages? It seems that `-devel` packages depend on the base ones already.

File: SPECS/pango/pango.spec [hunk 17]
```diff
@@ -1,58 +1,68 @@
...
+BuildRequires:  freetype
+BuildRequires:  glib-devel
+BuildRequires:  gobject-introspection-devel
+BuildRequires:  harfbuzz
+BuildRequires:  harfbuzz-devel
+BuildRequires:  libpng-devel
+BuildRequires:  meson
+BuildRequires:  pkg-config
+BuildRequires:  pkgconfig(fribidi)
+Requires:       harfbuzz-devel
```
Comment (MateuszMalisz) [inline SPECS/pango/pango.spec:24]: Why are the devel files needed during the runtime?

File: SPECS/pango/pango.spec [hunk 18]
```diff
@@ -1,58 +1,68 @@
...
+%post -p /sbin/ldconfig
+%postun -p /sbin/ldconfig
 
 %files
 %defattr(-,root,root)
 %license COPYING
 %{_bindir}/*
 %{_libdir}/*.so*
 %{_datadir}/*
+%{_libdir}/girepository-1.0/Pango-1.0.typelib
```
Comment (MateuszMalisz) [inline SPECS/pango/pango.spec:61]: Any reason why we're not using a wildcard in here?

File: SPECS/python3/python3.spec [hunk 19]
```diff
@@ -27,6 +27,7 @@ Requires:       openssl
 Requires:       python3-libs = %{version}-%{release}
 Requires:       readline
 Requires:       xz
+Provides:       python
```
Comment (MateuszMalisz) [inline SPECS/python3/python3.spec:30]: Why do we provide python explicitly now?

**Review by PawelWMS — APPROVED** [review]
